FIX PHP 8.1 support in MySQLiConnector::query errors

NightJar opened the pull request:
Issue https://github.com/silverstripe/silverstripe-framework/issues/10613

The default error reporting mode in PHP 8.1 has changed from using errors reported on the connection handle to throwing mysqli_sql_exception. query() makes no allowance for this, and functions up the call stack expect to catch
Silverstripe\ORM\Connect\DatabaseException instead - resulting in the MySQLi exception going all the way up to halt the system.

We can use a try, catch, and finally to retain backwards compatibility, no matter which setting (e.g. PHP version default) someone has enabled.

C.f. https://www.php.net/manual/en/migration81.incompatible.php#migration81.incompatible.mysqli

Commit:
FIX PHP 8.1 support in MySQLiConnector::query errors

The default error reporting mode in PHP 8.1 has changed from using
errors reported on the connection handle to throwing
mysqli_sql_exception. query() makes no allowance for this, and
functions up the call stack expect to catch
Silverstripe\ORM\Connect\DatabaseException instead - resulting in the
MySQLi exception going all the way up to halt the system.

We can use a try, catch, and finally to retain backwards compatibility,
no matter which setting (e.g. PHP version default) someone has enabled.

NightJar commented:
This fixes an [issue for any site with subsites](https://github.com/silverstripe/silverstripe-subsites/blob/2/src/Middleware/InitStateMiddleware.php#L45) installed being unable to successfully run a dev/build without an existing database (e.g. setting up for development).

michalkleiner commented:
@NightJar I thought we were testing on PHP 8.1 before, do you know if there's a test relating to and/or why it was not failing before?

NightJar commented:
> do you know if there's a test

There are no existing tests, this change set adds some.

> why it was not failing before?

Probably the same reason there's [no change to the `use` statements](https://github.com/silverstripe/silverstripe-framework/blob/4.11.14/src/ORM/Connect/MySQLiConnector.php#L62-L69) atop the file in this change set ;)

I presume because the most common code path is to use `prepareStatement` rather than `query`... OR even more simply that the calling codes purpose is to _not_ cause an error.

In the Subsites case this is an issue only because the code may run [before the database is ready](https://github.com/silverstripe/silverstripe-subsites/blob/ceaa915b77512e3f9d63344496e4245ef95d6a41/src/Middleware/InitStateMiddleware.php#L50) (e.g. first `dev/build` on a fresh dev environment).

michalkleiner reviewed (approved):
Looks ok to me but I'd prefer another set of @silverstripe/core-team eyes over it.

emteknetnz commented on tests/php/ORM/MySQLiConnectorTest.php:
```diff
@@ -2,32 +2,48 @@
             return $this->markTestSkipped('The test only relevant for MySQL');
         }
 
+        if ($charset) {
```
Moving the duplicate logic to this new private method is OK, except for the bit where the marking test as skipped if it's it's not running mysql - It makes it far less obvious could you please re-duplicate those bits and move those lines back to the individual tests

NightJar replied:
I'll put it in `setUp` instead. This makes it super obvious it runs before each test case.

emteknetnz reviewed (changes requested):
Thanks for doing this.

Looks fine, though small thing would like marking the tests as skipped moved back out of the new private method

Commit:
Move MySQLConnector test skip call into setUp()

As review feedback; marking the test as skipped in a private function
obfuscated where the call was happening and made it harder to skimread
the tests. Moving this into a setUp function makes it obvious the check
is run before each test case, and skipped if necessary.

MasonD commented:
preparedQuery actually still has this same problem. There’s a catch in prepareStatement but no catch around $statement->execute() in preparedQuery itself, encountering the same problem as query is having if the sql is syntactically correct but the execution causes an error (for example inserting/updating that causes unique contraint clashes)

emteknetnz reviewed (approved):
Looks good. I've re-targeted this PR to 4.13 since it's the oldest currently supported version.  I've run a fork on this PR (rebased on 4.13) against silverstripe/installer [here](https://github.com/emteknetnz/silverstripe-installer/actions/runs/5482285324) which went green
